pep-0705.rst

```diff
@@ -0,0 +1,259 @@
+
+Take for example the function ``movie_string`` defined above. Suppose its type signature is defined using a ``TypedDict``::
+
+    from typing import NotRequired, TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: NotRequired[int | None]
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+    def movie_string(movie: Movie) -> str:
+        if movie.get("year") is None:
+            return movie["name"]
+        else:
+            return f'{movie["name"]} ({movie["year"]})'
+
+Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::
+
+    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"
+
+This particular use case should be type-safe, but the type checker correctly stops the
+user from passing a ``MovieRecord`` into a ``Movie`` argument in the general case because
+the ``Movie`` class has mutator methods that could potentially allow the function to break
+the type constraints in ``MovieRecord`` by setting ``movie["year"] = None`` or ``del movie["year"]``.
+The problem disappears if we don't have mutator methods in ``Movie``. This could be achieved by defining an immutable interface using a :pep:`544` Protocol::
+
+    from typing import Literal, Protocol, overload
+
+    class Movie(Protocol):
+        @overload
+        def get(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def get(self, key: Literal["year"]) -> int | None: ...
+
+        @overload
+        def __getitem__(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def __getitem__(self, key: Literal["year"]) -> int | None: ...
+
+This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.
+
+The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above.
+In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
+Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.
+
+Specification
+=============
+
+The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:
+
+1. The runtime type of a TypedMapping object is not constrained to be a ``dict``
+2. no mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
+3. a class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
+4. subclasses can narrow field types, in the same manner as other protocols
```

Comment:
```suggestion
* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``
* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
* Subclasses can narrow field types, in the same manner as other protocols
```

* Use a bulleted rather than ordered list for items that aren't sequential and don't have a required, unambiguous order
* Use consistent and correct capitalization
* Add missing literal to one ``TypedMapping`` instance